src/main/scala/com.snowplowanalytics/sqs2kinesis/EventsStreamModule.scala

```diff
@@ -58,89 +55,133 @@ object EventsStreamModule {
+  case class ParsedMsg(original: Message, key: String, bytes: Array[Byte])
 
-    val sqsMsg2kinesisMsg: Message => KinesisKeyAndMsg =
-      msg => {
-        val decoded  = ByteBuffer.wrap(java.util.Base64.getDecoder.decode(msg.body))
+  /** A flow that base64-decodes sqs messages. It ignores messages if they cannot be parsed */
+  val sqsMsg2kinesisMsg: Flow[Message, ParsedMsg, NotUsed] =
+    Flow[Message]
+      .map { msg =>
+        logger.debug(s"Received message ${msg.messageId}")
+        val decoded  = java.util.Base64.getDecoder.decode(msg.body)
         val maybeKey = Option(msg.messageAttributes().get(KinesisKey)).map(_.stringValue())
         val key = maybeKey.getOrElse {
           val randomKey = UUID.randomUUID().toString
           logger.warn(s"Kinesis key for sqs message ${msg.messageId()} not found, random key generated: $randomKey")
           randomKey
         }
-        (key, decoded)
+        Some(ParsedMsg(msg, key, decoded))
+      }.recover {
+        case NonFatal(e) =>
+          logger.error("Error decoding sqs message. Message will be ignored and not acked.", e)
+          None
+      }.mapConcat(_.toSeq)
+
+  def toPutRecordReqEntry(msg: ParsedMsg): PutRecordsRequestEntry = {
+      new PutRecordsRequestEntry().withPartitionKey(msg.key).withData(ByteBuffer.wrap(msg.bytes))
+  }
+
+  /** A Flow that tries to send a batch of messages to kinesis. Any failures in the batch will be retried up to 5 times. */
+  def kinesisFlow(config: Sqs2KinesisConfig, kinesisClient: AmazonKinesisAsync): Flow[Vector[ParsedMsg], Message, NotUsed] = {
+
+    // The innter flow, which must be retried on error.
+    val inner = Flow[(Vector[ParsedMsg], Vector[ParsedMsg])]
+      .map { case (todo, complete) =>
+        val req = new PutRecordsRequest()
+          .withStreamName(config.kinesisStreamName)
+          .withRecords(todo.map(toPutRecordReqEntry).asJavaCollection)
+        try {
+          val results = kinesisClient.putRecords(req).getRecords.asScala.toList
+          val (successes, failures) = todo.zip(results).partition(_._2.getErrorMessage == null)
+          (complete ++ successes.map(_._1), failures.map(_._1))
+        } catch {
+          case NonFatal(e) =>
+            logger.error("Writing to kinesis failed with error", e)
+            (complete, todo)
+        }
       }
 
-    val toPutRecordReqEntry: KinesisKeyAndMsg => PutRecordsRequestEntry = {
-      case (key, data) =>
-        new PutRecordsRequestEntry().withPartitionKey(key).withData(data)
-    }
+    Flow[Vector[ParsedMsg]]
+      .map(_ -> Vector.empty)
+      .via(RetryFlow.withBackoff(500.milli, 1.second, 0.0, 5, inner) {
+        case (_, (_, Vector())) =>
+          None
+        case (_, (successes, failures)) =>
+          if (failures.nonEmpty)
+            logger.error(s"Got ${failures.size} failures and ${successes.size} successes writing to kinesis. Failures will be retried.")
+          Some((failures, successes))
```

Comment:
By returning `Some` we are saying that something needs to be retried.  The tuple we return in the `Some` is `(failures, successes)` and those values are then passed back to the inner flow as the `(todo, complete)` input parameters.

[This](https://doc.akka.io/docs/akka/current/stream/operators/RetryFlow/withBackoff.html) is the documentation I am following.

Reply:
Ah, I see I made a mistake - but is not quite as you say.  If failures *is* empty then the pattern on line 166 matches `case (_, (_, Vector()))` and we do indeed return `None`, as you say we should.

So my mistake is the if statement `if (failures.nonEmpty)` is redundant, because this will always be true.